### ISSUE TITLE
oneflow backend yolov5 性能调试

### ISSUE THREAD

**Issue by BBuf** [Oneflow-Inc/one-yolov5#46]
yolov5在训练时的流程大概如下：

```python
for i in range(epochs):
   for j in range(data):
     input, target = data[i]
     out = forward(input)
     loss = compute_loss(out, target)
     loss.backward()
     optimizer.step()
  
  cal_map()
```

把上面的for j in range(data)里面包含的内容叫作 train。

对上述关键步骤用nsys分别对pytorch yolov5和oneflow yolov5进行埋点，以相同的配置运行2个epoch记录结果。观察第二个epoch的结果，发现对于train的部分，pytorch的耗时在1.172s，而oneflow的耗时在800ms左右。此数据说明了对于backbone部分的前向，反向，以及参数更新oneflow都明显优于pytorch。

但是在cal_map部分oneflow的耗时为7.5s，而pytorch只需要272ms，此差距十分巨大，也是变慢的主要原因。进一步，将cal_map部分的主要流程梳理如下：

```python
for i in range(len(val_data)):
  im, target = val_data[i]:
  out = model(im)
  loss = compute_loss(out, target)
  out = nms(out)
  record_metrics()

cal_metrics()
```

在上述关键位置分别埋点，获得的nsys结果如下：

pytorch:

![图片](https://user-images.githubusercontent.com/35585791/194743447-2e2ebd14-a827-41a5-a5ee-58ef033951b1.png)

oneflow:

![图片](https://user-images.githubusercontent.com/35585791/194743503-98c91ca7-c586-4dbe-9133-eaac1571ac83.png)

上述伪代码中的关键操作，在pytorch中均在20ms左右完成，但在oneflow里几乎都是几百ms到1s多之间。并且观察这时候的cuda kernel，我们发现pytorch的cuda kernel还是断续在执行的，而oneflow里则几乎没有执行cuda kernel，而是大量的copy d2h操作。接下来需要定位一下这些d2h操作是如何引入的，找到cal_map这部分的性能瓶颈。

**Comment by BBuf** [comment]
定位到上述问题产生的原因在于训练脚本里面每个epoch做计算map时，pytorch会关闭plot_image这个选项，而我们的脚本是默认打开的。对齐这个参数之后我们的整体性能和pytorch可以持平，为了进一步分析性能这里把forward, compute_loss, loss.backward()部分统称为训练过程，把计算map部分统称为后处理过程。以下分别为pytorch和oneflow的nsys结果：

![图片](https://user-images.githubusercontent.com/35585791/194789839-dbb08735-71a6-4dbe-9888-7b9574d272fb.png)

![图片](https://user-images.githubusercontent.com/35585791/194789874-a5f4bcbf-33c2-4f0d-9493-050731aabecb.png)

可以看到pytorch训练过程和后处理过程的耗时分别为：1.172s 和 278ms。
              oneflow 训练过程和后处理过程的耗时分别为：795ms 和 527ms。

所以oneflow这边在训练过程有很大的速度优势，但是在后处理部分却比pytorch慢了2倍，这就导致最终的整体性能和pytorch只能持平。

继续分析，发现这个速度差距主要来源于后处理过程中的nms过程，我们的耗时平均在70ms，而pytorch只有10-20ms。由于后处理过程在不断取数据进行nms，所以这个速度差距随着数据增多就越来越大了。

oneflow nms：

![图片](https://user-images.githubusercontent.com/35585791/194790280-d852b250-1e7c-4598-8d68-8c0d515f8c86.png)

pytorch nms：

![图片](https://user-images.githubusercontent.com/35585791/194790309-0b2f46bb-0e31-44c5-af08-73c08c1b2dcc.png)

可以看到pytorch的nms过程cuda kernel排得非常精密，我们则是非常稀疏并且有很多大的空隙。

继续分析，发现在nms过程中有很多次nonzero（存在同步）和slice_update，这两个操作应该是速度慢的重要原因。

<img width="686" alt="图片" src="https://user-images.githubusercontent.com/35585791/194790477-72ff7c2f-046b-4917-9520-bdf6038b07db.png">

**Comment by lixinqi** [comment]
nonzero比pytorch慢一倍，这确实是oneflow目前的表现。

**Comment by BBuf** [comment]
通过上述实验的发现，我把nms这部分单独拆成了一个脚本，然后配置参数和yolov5工程里面的完全一样。脚本如下：

<details>
  <summary>PyTorch脚本</summary>
  <pre><code> 
     import time
import torchvision
import torch

def xywh2xyxy(x):
    # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = x.clone() if isinstance(x, torch.Tensor) else np.copy(x)
    y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
    y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
    y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
    y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
    return y

def box_area(box):
    # box = xyxy(4,n)
    return (box[2] - box[0]) * (box[3] - box[1])

def box_iou(box1, box2, eps=1e-7):
    # inter(N,M) = (rb(N,M,2) - lt(N,M,2)).clamp(0).prod(2)
    (a1, a2), (b1, b2) = box1[:, None].chunk(2, 2), box2.chunk(2, 1)
    inter = (torch.min(a2, b2) - torch.max(a1, b1)).clamp(0).prod(2)

    # IoU = inter / (area1 + area2 - inter)
    return inter / (box_area(box1.T)[:, None] + box_area(box2.T) - inter + eps)


def non_max_suppression(
    prediction,
    conf_thres=0.25,
    iou_thres=0.45,
    classes=None,
    agnostic=False,
    multi_label=False,
    labels=(),
    max_det=300,
):
    """Non-Maximum Suppression (NMS) on inference results to reject overlapping bounding boxes

    Returns:
         list of detections, on (n,6) tensor per image [xyxy, conf, cls]
    """
    bs = prediction.shape[0]  # batch size
    nc = prediction.shape[2] - 5  # number of classes
    xc = prediction[..., 4] > conf_thres  # candidates

    # Checks
    assert 0 <= conf_thres <= 1, f"Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0"
    assert 0 <= iou_thres <= 1, f"Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0"

    # Settings
    # min_wh = 2  # (pixels) minimum box width and height
    max_wh = 7680  # (pixels) maximum box width and height
    max_nms = 30000  # maximum number of boxes into oneflowvision.ops.nms()
    time_limit = 0.3 + 0.03 * bs  # seconds to quit after
    redundant = True  # require redundant detections
    multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)
    merge = False  # use merge-NMS

    t = time.time()
    output = [torch.zeros((0, 6), device=prediction.device)] * bs
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]):
            lb = labels[xi]
            v = torch.zeros((len(lb), nc + 5), device=x.device)
            v[:, :4] = lb[:, 1:5]  # box
            v[:, 4] = 1.0  # conf
            v[range(len(lb)), lb[:, 0].long() + 5] = 1.0  # cls
            x = torch.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Compute conf
        x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

        # Box (center x, center y, width, height) to (x1, y1, x2, y2)
        box = xywh2xyxy(x[:, :4])

        # Detections matrix nx6 (xyxy, conf, cls)
        if multi_label:
            i, j = (x[:, 5:] > conf_thres).nonzero(as_tuple=False).T
            x = torch.cat((box[i], x[i, j + 5, None], j[:, None].float()), 1)
        else:  # best class only
            conf, j = x[:, 5:].max(1, keepdim=True)
            x = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes is not None:
            x = x[(x[:, 5:6] == torch.tensor(classes, device=x.device)).any(1)]

        # Apply finite constraint
        # if not torch.isfinite(x).all():
        #     x = x[torch.isfinite(x).all(1)]

        # Check shape
        n = x.shape[0]  # number of boxes
        if not n:  # no boxes
            continue
        elif n > max_nms:  # excess boxes
            x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        if i.shape[0] > max_det:  # limit detections
            i = i[:max_det]
        if merge and (1 < n < 3e3):  # Merge NMS (boxes merged using weighted mean)
            # update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
            iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
            weights = iou * scores[None]  # box weights
            x[i, :4] = torch.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
            if redundant:
                i = i[iou.sum(1) > 1]  # require redundancy

        output[xi] = x[i]
        if (time.time() - t) > time_limit:
            break  # time limit exceeded

    return output

if __name__ == "__main__":
    torch.cuda.nvtx.range_push('loop begin')
    for i in range(20):
        torch.cuda.nvtx.range_push('to cuda')
        input = torch.ones(32, 18522, 85).to("cuda")
        torch.cuda.nvtx.range_pop()
        torch.cuda.nvtx.range_push('non_max_suppression')
        output = non_max_suppression(input, conf_thres=0.001, iou_thres=0.6, labels=[], multi_label=True, agnostic=False)
        torch.cuda.nvtx.range_pop()
        torch.cuda.nvtx.range_push('synchronize')
        for j in range(len(output)):
            out = output[j].cpu().detach().numpy()
        torch.cuda.nvtx.range_pop()
    torch.cuda.nvtx.range_pop()

  </code></pre>
</details>


<details>
  <summary>oneflow 脚本</summary>
  <pre><code> 
     import time
import torchvision
import oneflow as flow

def xywh2xyxy(x):
    # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = x.clone() if isinstance(x, flow.Tensor) else np.copy(x)
    y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
    y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
    y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
    y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
    return y

def box_area(box):
    # box = xyxy(4,n)
    return (box[2] - box[0]) * (box[3] - box[1])

def box_iou(box1, box2, eps=1e-7):
    # inter(N,M) = (rb(N,M,2) - lt(N,M,2)).clamp(0).prod(2)
    (a1, a2), (b1, b2) = box1[:, None].chunk(2, 2), box2.chunk(2, 1)
    inter = (flow.min(a2, b2) - flow.max(a1, b1)).clamp(0).prod(2)

    # IoU = inter / (area1 + area2 - inter)
    return inter / (box_area(box1.T)[:, None] + box_area(box2.T) - inter + eps)


def non_max_suppression(
    prediction,
    conf_thres=0.25,
    iou_thres=0.45,
    classes=None,
    agnostic=False,
    multi_label=False,
    labels=(),
    max_det=300,
):
    """Non-Maximum Suppression (NMS) on inference results to reject overlapping bounding boxes

    Returns:
         list of detections, on (n,6) tensor per image [xyxy, conf, cls]
    """
    bs = prediction.shape[0]  # batch size
    nc = prediction.shape[2] - 5  # number of classes
    xc = prediction[..., 4] > conf_thres  # candidates

    # Checks
    assert 0 <= conf_thres <= 1, f"Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0"
    assert 0 <= iou_thres <= 1, f"Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0"

    # Settings
    # min_wh = 2  # (pixels) minimum box width and height
    max_wh = 7680  # (pixels) maximum box width and height
    max_nms = 30000  # maximum number of boxes into oneflowvision.ops.nms()
    time_limit = 0.3 + 0.03 * bs  # seconds to quit after
    redundant = True  # require redundant detections
    multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)
    merge = False  # use merge-NMS

    t = time.time()
    output = [flow.zeros((0, 6), device=prediction.device)] * bs
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]):
            lb = labels[xi]
            v = flow.zeros((len(lb), nc + 5), device=x.device)
            v[:, :4] = lb[:, 1:5]  # box
            v[:, 4] = 1.0  # conf
            v[range(len(lb)), lb[:, 0].long() + 5] = 1.0  # cls
            x = flow.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Compute conf
        x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

        # Box (center x, center y, width, height) to (x1, y1, x2, y2)
        box = xywh2xyxy(x[:, :4])

        # Detections matrix nx6 (xyxy, conf, cls)
        if multi_label:
            i, j = (x[:, 5:] > conf_thres).nonzero(as_tuple=False).T
            x = flow.cat((box[i], x[i, j + 5, None], j[:, None].float()), 1)
        else:  # best class only
            conf, j = x[:, 5:].max(1, keepdim=True)
            x = flow.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes is not None:
            x = x[(x[:, 5:6] == flow.tensor(classes, device=x.device)).any(1)]

        # Apply finite constraint
        # if not flow.isfinite(x).all():
        #     x = x[flow.isfinite(x).all(1)]

        # Check shape
        n = x.shape[0]  # number of boxes
        if not n:  # no boxes
            continue
        elif n > max_nms:  # excess boxes
            x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        try:
            i = flow.nms(boxes, scores, iou_thres)  # NMS
        except:
            i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        if i.shape[0] > max_det:  # limit detections
            i = i[:max_det]
        if merge and (1 < n < 3e3):  # Merge NMS (boxes merged using weighted mean)
            # update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
            iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
            weights = iou * scores[None]  # box weights
            x[i, :4] = flow.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
            if redundant:
                i = i[iou.sum(1) > 1]  # require redundancy

        output[xi] = x[i]
        if (time.time() - t) > time_limit:
            break  # time limit exceeded

    return output

if __name__ == "__main__":
    flow._oneflow_internal.profiler.RangePush('loop begin')
    for i in range(20):
        flow._oneflow_internal.profiler.RangePush('to cuda')
        input = flow.ones(32, 18522, 85).to("cuda")
        flow._oneflow_internal.profiler.RangePop()
        flow._oneflow_internal.profiler.RangePush('non_max_suppression')
        output = non_max_suppression(input, conf_thres=0.001, iou_thres=0.6, labels=[], multi_label=True, agnostic=False)
        flow._oneflow_internal.profiler.RangePop()
        flow._oneflow_internal.profiler.RangePush('synchronize')
        for j in range(len(output)):
            out = output[j].cpu().detach().numpy()
        flow._oneflow_internal.profiler.RangePop()
    flow._oneflow_internal.profiler.RangePop()

  </code></pre>
</details>


然后oneflow的nsys结果如下：

![图片](https://user-images.githubusercontent.com/35585791/194810895-c94f5753-4e01-4bf7-bdc1-d43a2adf637f.png)

pytorch的nsys结果如下：

![图片](https://user-images.githubusercontent.com/35585791/194810986-981448d3-bc09-4d55-82ff-efee68a11ea5.png)

非常奇怪的是这个nsys结果中pytorch的主要操作都是cudaMemcpyAsync，而oneflow的执行结果看起来就非常正常。

@lixinqi 对这个问题有什么想法么？我稍后上传一下nsys到oss。

**Comment by lixinqi** [comment]
这里的cudaMemcpyAsync是h2d还是d2h呢？

**Comment by BBuf** [comment]
> 这里的cudaMemcpyAsync是h2d还是d2h呢？

d2h

**Comment by liujuncheng** [comment]
<img width="880" alt="image" src="https://user-images.githubusercontent.com/3319142/195560003-d32c2b4c-2927-4e76-9ec4-061862aa5446.png">

这个 copy 的目的地址不是pinned memory

**Comment by BBuf** [comment]
> <img alt="image" width="880" src="https://user-images.githubusercontent.com/3319142/195560003-d32c2b4c-2927-4e76-9ec4-061862aa5446.png">
> 
> 这个 copy 的目的地址不是pinned memory

哦哦，应该是没有设置non_blocking的原因？我试试。

**Comment by BBuf** [comment]
> <img alt="image" width="880" src="https://user-images.githubusercontent.com/3319142/195560003-d32c2b4c-2927-4e76-9ec4-061862aa5446.png">
> 
> 这个 copy 的目的地址不是pinned memory

我在pytorch的脚本里面to("cuda")的时候设置了non_blocking=True，发现这个问题还是存在。
